sklearn/feature_extraction/text.py

```diff
@@ -133,12 +133,24 @@ def _word_ngrams(self, tokens, stop_words=None):
         min_n, max_n = self.ngram_range
         if max_n != 1:
             original_tokens = tokens
-            tokens = []
+            if min_n == 1:
+                # no need to do any slicing for unigrams
+                # just iterate through the original tokens
+                tokens = list(original_tokens)
+                min_n += 1
+            else:
+                tokens = []
+
             n_original_tokens = len(original_tokens)
+
+            # bind method outside of loop to reduce overhead
+            tokens_append = tokens.append
+            space_join = " ".join
+
             for n in xrange(min_n,
                             min(max_n + 1, n_original_tokens + 1)):
                 for i in xrange(n_original_tokens - n + 1):
-                    tokens.append(" ".join(original_tokens[i: i + n]))
+                    tokens_append(space_join(original_tokens[i: i + n]))
```

Comment:
Actually there seem to be an even simpler way, using the following

```python
import itertools
def iter_window(seq, n):
    l = list(seq[:n])
    append = l.append
    for item in itertools.islice(seq, n-1, len(seq)):
        yield tuple(l)
        l.pop(0)
        l.append(item)
    yield tuple(l)
```

and then replacing the inner loop by

```python
for n in range(a,b):
            for _tks in iter_window(original_tokens,n, c-1):
                tokens_append(
                    space_join(
                        _tks
                    )
                )
```

I get a ~2+ speedup (increasing with lenght of ngrams)  and the sliding window implementation can likely be made much more efficient using a [dequeue](https://docs.python.org/2/library/collections.html#collections.deque) maybe.